[c10d] Fix data corruption bug after CUDAEventCache is enabled

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138059
* #138048
* __->__ #138040

Here is why we see using `CUDAEventCache` cause crash and data corruption.
1. The deleter is doing its job and append the job the stack.
2. In create, instead of getting a reference, we are getting a copy of eventsArray_[i] (which is a std::vector). This is bad because we didn't really remove the element from the stack. While we thought we already pop up the last one from the stack, but it turns out the last one is still in the stack; we end up reusing the same event again and again. What's worse, since we keep adding new events to the stack, this will eventually explode the stack and a crash happens. 

Fix is easy, just get a reference. Local torchtitan run see a non-Nan loss.

Also we want to use a deque instead of a stack, and refactor the code a bit to make it more readable. (in a separate PR)



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[c10d] Fix the bug inside CUDAEventCache::create

[ghstack-poisoned]

kwen2501 reviewed (approved):
Good finding!
I guess in a next PR we can turn on the cache?

fduwjj committed:
Update on "[c10d] Fix data corruption when CUDAEventCache is enabled"


Here is why we see using `CUDAEventCache` cause crash and data corruption.
1. The deleter is doing its job and append the job the stack.
2. In create, instead of getting a reference, we are getting a copy of eventsArray_[i] (which is a std::vector). This is bad because we didn't really remove the element from the stack. While we thought we already pop up the last one from the stack, but it turns out the last one is still in the stack; we end up reusing the same event again and again. What's worse, since we keep adding new events to the stack, this will eventually explode the stack and a crash happens. 

Fix is easy, just get a reference. Local torchtitan run see a non-Nan loss.
Also we want to use a deque instead of a stack, and refactor the code a bit to make it more readable.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 


[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -764,16 +764,15 @@ std::shared_ptr<at::cuda::CUDAEvent> ProcessGroupNCCL::CUDAEventCache::create(
   {
     std::lock_guard<std::mutex> lock(cacheMutex_);
-    auto events = eventsArray_[timing ? 1 : 0];
+    auto& events = eventsArray_[timing ? 1 : 0];
```
so there are 2 vectors of events, one for timing enabled and one for timing disabled

fduwjj replied:
yes. And each one is a std::vector.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -764,16 +764,15 @@ std::shared_ptr<at::cuda::CUDAEvent> ProcessGroupNCCL::CUDAEventCache::create(
-      event = events.back();
-      events.pop_back();
+      event = events.front();
+      events.pop_front();
```
any reason to prefer one order over the other?  (why did you switch from stack to fifo?)

fduwjj replied:
because fifo will return the oldest event first which makes it safter?

fduwjj committed:
Update on "[c10d] Fix data corruption bug after CUDAEventCache is enabled"


Here is why we see using `CUDAEventCache` cause crash and data corruption.
1. The deleter is doing its job and append the job the stack.
2. In create, instead of getting a reference, we are getting a copy of eventsArray_[i] (which is a std::vector). This is bad because we didn't really remove the element from the stack. While we thought we already pop up the last one from the stack, but it turns out the last one is still in the stack; we end up reusing the same event again and again. What's worse, since we keep adding new events to the stack, this will eventually explode the stack and a crash happens. 

Fix is easy, just get a reference. Local torchtitan run see a non-Nan loss.

Also we want to use a deque instead of a stack, and refactor the code a bit to make it more readable. (in a separate PR)



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o 


[ghstack-poisoned]

fduwjj commented:
Separate fix and refactor so that it is better for reviewers.

fduwjj commented:
@pytorchbot merge

wconstab commented:
You should probably add a unit test for the cache, that would have caught this.

fduwjj commented:
@wconstab that is a good point, maybe do you have any suggestions on how I can capture this in the unit test added in https://github.com/pytorch/pytorch/pull/137798? I tried to repro the error in unit test but no, it still pass. I even tried out composability test. My guess is that this bug needs a large scale to hit?

wconstab commented:
What I meant by unit test is literally a unit test of the cache. Put 3 events in, then get them back out and make sure you get the expected events and the cache is empty at the end.

fduwjj commented:
@wconstab Ahh, I see. Sure, let me write one.
